Global queue

gleb-cloudius opened the pull request:
**Please replace this line with justification for the backport/\* labels added to this PR**

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17451/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17451/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17451/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_truncate_with_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5340/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16592/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8508/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17451/)
#### Build Details:
- Duration: 7 hr 42 min
- Builder: spider2.cloudius-systems.com

gleb-cloudius commented:
@patjed41 ping

patjed41 reviewed (commented):
Looks good, I mostly left small comments.

I have some general questions:
1. There is a new truncate test, but don't we need a more general test for the queue? For example, we could send some global requests (perhaps a random sequence of 5-10 requests), control their execution one by one (with an error injection), and verify that each request has an expected effect.
2. How do we know that we don't introduce a regression for the old code (with the new cluster feature not supported)? All tests start a new cluster, so they use the new queue from the start. One idea I have is to run CI on a custom branch that has all the changes from this PR, but also has the new cluster feature disabled. We can run such a CI job in byo.
3. Is it possible that a node inserts a global request when the new cluster feature is not supported, but the topology coordinator handles the request when the feature is supported? I think it's impossible because:
- `raft_check_and_repair_cdc_streams` looks at the current topology state (read barrier),
- in `handle_topology_transition`, `handle_global_request` is above `enable_features`.

We would have to consider some complicated cases if it was possible.
4. What about https://github.com/scylladb/scylladb/issues/16822#issuecomment-2590115554, is it unneeded?

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -4994,27 +4994,27 @@ future<> storage_service::raft_check_and_repair_cdc_streams() {
         if (_topology_state_machine._topology.global_request_id) {
```
"topology coordinator:" is about subsystem changed, not about a file. Git shows what files were changed anyway.

gleb-cloudius replied:
> Github shows this comment in other commits sadly.

The solution is to not use github, but release engineering broke CI for patches.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -4990,9 +4991,22 @@ future<> storage_service::raft_check_and_repair_cdc_streams() {
         }
 
+        if (_topology_state_machine._topology.global_request_id) {
+            request_id = *_topology_state_machine._topology.global_request_id;
```
Not sure it is needed. In truncate it was used in a test.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -260,8 +260,8 @@ schema_ptr system_keyspace::topology() {
             .with_column("new_cdc_generation_data_uuid", timeuuid_type, column_kind::static_column)
-            .with_column("new_keyspace_rf_change_ks_name", utf8_type, column_kind::static_column)
-            .with_column("new_keyspace_rf_change_data", map_type_impl::get_instance(utf8_type, utf8_type, false), column_kind::static_column)
+            .with_column("new_keyspace_rf_change_ks_name", utf8_type, column_kind::static_column) // deprecated
+            .with_column("new_keyspace_rf_change_data", map_type_impl::get_instance(utf8_type, utf8_type, false), column_kind::static_column) // deprecated
```
I do not think so. Only by moving to topology_v2 or something.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -1071,6 +1071,7 @@ class storage_proxy::remote {
+            if (!_sp._features.topology_global_request_queue) {
             // Check if we already have a truncate queued for the same table. This can happen when a truncate has timed out
             // and the client retried by issuing the same truncate again. In this case, instead of failing the request with
             // an "Another global topology request is ongoing" error, we can wait for the already queued request to complete.
```
What is the problem? All the tricks here were because we could not queue new request but did not want to fail truncate for the same node. Now we can.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17581/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17581/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17581/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_truncate_with_tablets
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16686/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8603/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5435/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17581/)
#### Build Details:
- Duration: 8 hr 7 min
- Builder: spider2.cloudius-systems.com

avikivity commented on db/system_keyspace.cc:
```diff
@@ -273,6 +273,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("session", uuid_type, column_kind::static_column)
             .with_column("tablet_balancing_enabled", boolean_type, column_kind::static_column)
             .with_column("upgrade_state", utf8_type, column_kind::static_column)
+            .with_column("global_requests", set_type_impl::get_instance(timeuuid_type, true), column_kind::static_column)
```
static + set =~ table with clustering key

The advantage of a new table with a clustering key is that you can easily add columns as needed.

So how about system.topology_request_queue?

gleb-cloudius replied:
> static + set =~ table with clustering key
> 
> The advantage of a new table with a clustering key is that you can easily add columns as needed.
> 

Why would I need to? Actual requests parameters are in a separate table where new columns can be added as needed.

> So how about system.topology_request_queue?

I do not see and advantage, but see a lot of complications since a topology state will be spread over several tables.

avikivity replied:
> > static + set =~ table with clustering key
> > The advantage of a new table with a clustering key is that you can easily add columns as needed.
> 
> Why would I need to? Actual requests parameters are in a separate table where new columns can be added as needed.


Then why do we need this column? Use that table directly.

> 
> > So how about system.topology_request_queue?
> 
> I do not see and advantage, but see a lot of complications since a topology state will be spread over several tables.

Isn't this column spreading the data over yet another table?

gleb-cloudius replied:
> > > static + set =~ table with clustering key
> > > The advantage of a new table with a clustering key is that you can easily add columns as needed.
> > 
> > 
> > Why would I need to? Actual requests parameters are in a separate table where new columns can be added as needed.
> 
> Then why do we need this column? Use that table directly.

Because it is a list of pending requests. Requests table contains all the requests, those that were completed as well. We do not want to read it all and scat it to find pending requests.

> 
> > > So how about system.topology_request_queue?
> > 
> > 
> > I do not see and advantage, but see a lot of complications since a topology state will be spread over several tables.
> 
> Isn't this column spreading the data over yet another table?

Another table already exists and is used during request processing. "Pending requests" is a state machine property, Before that there could have been only one so it was kept in `global_topology_request_id` in the topology table.

avikivity replied:
> > > > static + set =~ table with clustering key
> > > > The advantage of a new table with a clustering key is that you can easily add columns as needed.
> > > 
> > > 
> > > Why would I need to? Actual requests parameters are in a separate table where new columns can be added as needed.
> > 
> > 
> > Then why do we need this column? Use that table directly.
> 
> Because it is a list of pending requests. Requests table contains all the requests, those that were completed as well. We do not want to read it all and scat it to find pending requests.

Why do we need completed requests?

We could move them to another partition, if we do need them.


> 
> > > > So how about system.topology_request_queue?
> > > 
> > > 
> > > I do not see and advantage, but see a lot of complications since a topology state will be spread over several tables.
> > 
> > 
> > Isn't this column spreading the data over yet another table?
> 
> Another table already exists and is used during request processing. "Pending requests" is a state machine property, Before that there could have been only one so it was kept in `global_topology_request_id` in the topology table.

It's better to avoid duplication, and to avoid collection data types if we don't have a clear bound on their size.

gleb-cloudius replied:
> > > > > static + set =~ table with clustering key
> > > > > The advantage of a new table with a clustering key is that you can easily add columns as needed.
> > > > 
> > > > 
> > > > Why would I need to? Actual requests parameters are in a separate table where new columns can be added as needed.
> > > 
> > > 
> > > Then why do we need this column? Use that table directly.
> > 
> > 
> > Because it is a list of pending requests. Requests table contains all the requests, those that were completed as well. We do not want to read it all and scat it to find pending requests.
> 
> Why do we need completed requests?

First because we signal completion there. There can be multiple waiters. And second it provides a nice history for debugging purposes. The table is TTLed (for 30 days IIRC).

> 
> We could move them to another partition, if we do need them.

Each request entry is in a separate partition.

> 
> > > > > So how about system.topology_request_queue?
> > > > 
> > > > 
> > > > I do not see and advantage, but see a lot of complications since a topology state will be spread over several tables.
> > > 
> > > 
> > > Isn't this column spreading the data over yet another table?
> > 
> > 
> > Another table already exists and is used during request processing. "Pending requests" is a state machine property, Before that there could have been only one so it was kept in `global_topology_request_id` in the topology table.
> 
> It's better to avoid duplication, and to avoid collection data types if we don't have a clear bound on their size.

There is no duplication. This this the only place that holds pending requests queue. A bound can be enforced if needed.

gleb-cloudius commented:
> What about some tests?

There is a test for multiple truncates. This test the queue machinery. It would be nice to extend random test to allow submitting multiple commands, but I will need help for that from the test author.

gleb-cloudius commented:
> I see absolutely no documentation. The new persistent state and the state machine must be documented.

I will document new field in docs/dev/topology-over-raft.md, good point. But there is no change to the state machine in the series. It only removes the limitation that, as far as I see, is not even mentioned in the doc.

gleb-cloudius commented:
Rebased and documented new cell.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -1071,6 +1071,7 @@ class storage_proxy::remote {
             const table_id table_id = _sp.local_db().find_uuid(ks_name, cf_name);
 
+            if (!_sp._features.topology_global_request_queue) {
             // Check if we already have a truncate queued for the same table. This can happen when a truncate has timed out
```
I see. Yes we can simplify here. Will make the patch more complicated. I'll add a cleanup patch.

gleb-cloudius commented:
Added patch to simplify non queue enable truncate code after enabling the queue. Added assert that if a request queue is not empty the feature it is enabled. @patjed41

bhalevy commented:
> Added patch to simplify non queue enable truncate code after enabling the queue. Added assert that if a request queue is not empty the feature it is enabled. @patjed41

Is the new patch part of this PR?
@gleb-cloudius maybe you didn't push the branch?

patjed41 reviewed (approved):
If a node inserts a global request when the new cluster feature is not supported, but the topology coordinator handles the request when the feature is supported, everything should be fine:
- the topology coordinator can take the request from `global_request` even if the feature is supported,
- the node that inserted the request will wait for the request completion in the old way, which is different for `cleanup` and `check_and_repair_cdc_streams` (the node won't look at `topology_request`). I don't see a problem with it.

patjed41 commented:
queued
